text/0225-configuration.md

```diff
@@ -0,0 +1,411 @@
+* Broadly supported format. Ideally, the information encoded in the file can be decoded using native tools for all OpenTelemetry implementation languages. However, it must be possible for languages that do not natively support an encoding format to write their own parsers.
+* The configuration format must support structured data. At the minimum arrays and associative arrays.
+* The format must support at least null, boolean, string, double precision floating point (IEEE 754-1985), or signed 64 bit integer value types.
+* Custom span processors, exporters, samplers, or other user defined extension components can be configured using this format.
+* Configure SDK, but also configure instrumentation.
+* Must offer stability guarantees while supporting evolution
+* The structure of the configuration can be validated via a schema.
+* Support environment variable substitution to give users the option to avoid storing secrets in these files.
+
+## Internal details
+
+The schema for OpenTelemetry configuration is to be published in a repository to allow language implementations to leverage that definition to automatically generate code and/or validate end-user configuration. This will ensure that all implementations provide a consistent experience for any version of the schema they support. An example of such a proposed schema is available [here](https://github.com/MrAlias/otel-schema/tree/main/json_schema/schema).
+
+The working group proposes the use of [JSON Schema](https://json-schema.org/) as the language to define the schema. It provides:
+
+* support for client-side validation
+* code generation
+* broad support across languages
+
+In order to provide a minimal API surface area, implementations *MUST* support the following methods.
+
+### Configure(config)
+
+An API called `Configure` receives a configuration object. This method then applies the configuration object's details to the SDK. This method specifically applies the configuration object to allow for multiple configuration format providers to be supported in the future. This OTEP describes two such providers in a file and data structure formats below, but remote file formats *MAY* be implemented in the future.
+
+### ParseAndValidateConfigurationFromFile(filepath, format) -> config
+
+An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. YAML *SHOULD* be supported, but JSON, which is a strict subset of YAML, *MAY* be supported if a language ecosystem has good reason. The method returns a `Configuration` model that has been validated. This API *MAY* return an error or raise an exception, whichever is idiomatic to the implementation for the following reasons:
```

Comment:
Are there examples of language ecosystems that don't support yaml?  Having yaml universally supported would be nice, but i'm guessing this is here because one such language came up during review.

Reply:
The issue isn't lack of yaml support but that some languages have builtin json support (Javascript, Go) and it makes sense to not require them, JS in particular, to include an additional dependency.

Reply:
Tip: include this motivation into the text directly.

Reply:
(JS _could_ offer yaml as well through an _additional and upcoming_ library)

Reply:
Included an example where json support may be preferable.

Reply:
To be clear, JS doesn't support YAML out of the box but the ecosystem has yaml parsers just like any other language. There are plenty of examples of JS libraries which use YAML for configuration.

That said, I would definitely prefer to use language builtins like JSON if possible. Thus far the JS SDK has been able to keep external dependencies to a minimum and I'd like to keep it that way.

It might actually be beneficial to allow arbitrary file types using some pluggable system. This could be accomplished easily using dynamic requires based on the configuration file extension in node for example.